pkg/destroy/aws/aws.go

```diff
@@ -165,132 +162,322 @@ func (o *ClusterUninstaller) RunWithContext(ctx context.Context) error {
 		logger:  o.Logger,
 	}
 
-	deleted, err := terminateEC2InstancesByTags(ctx, ec2.New(awsSession), iamClient, o.Filters, o.Logger)
+	// Get the initial resources to delete, so that they can be returned if the context is canceled while terminating
+	// instances.
+	deleted := sets.NewString()
+	resourcesToDelete, tagClientsWithResources, err := o.findResourcesToDelete(ctx, tagClients, iamClient, iamRoleSearch, iamUserSearch, deleted)
 	if err != nil {
-		return err
+		o.Logger.WithError(err).Info("error while finding resources to delete")
+		if err := ctx.Err(); err != nil {
+			return resourcesToDelete.UnsortedList(), err
+		}
 	}
 
 	tracker := new(errorTracker)
-	tagClientStack := append([]*resourcegroupstaggingapi.ResourceGroupsTaggingAPI(nil), tagClients...)
+
+	// Terminate EC2 instances.
+	ec2Client := ec2.New(awsSession)
 	err = wait.PollImmediateUntil(
 		time.Second*10,
 		func() (done bool, err error) {
-			var loopError error
-			nextTagClients := tagClients[:0]
-			for _, tagClient := range tagClientStack {
-				matched := false
-				for _, filter := range o.Filters {
-					o.Logger.Debugf("search for and delete matching resources by tag in %s matching %#+v", tagClientNames[tagClient], filter)
-					tagFilters := make([]*resourcegroupstaggingapi.TagFilter, 0, len(filter))
-					for key, value := range filter {
-						tagFilters = append(tagFilters, &resourcegroupstaggingapi.TagFilter{
-							Key:    aws.String(key),
-							Values: []*string{aws.String(value)},
-						})
-					}
-					err = tagClient.GetResourcesPagesWithContext(
-						ctx,
-						&resourcegroupstaggingapi.GetResourcesInput{TagFilters: tagFilters},
-						func(results *resourcegroupstaggingapi.GetResourcesOutput, lastPage bool) bool {
-							for _, resource := range results.ResourceTagMappingList {
-								arnString := *resource.ResourceARN
-								if _, ok := deleted[arnString]; !ok {
-									arnLogger := o.Logger.WithField("arn", arnString)
-									matched = true
-									parsed, err := arn.Parse(arnString)
-									if err != nil {
-										arnLogger.Debug(err)
-										continue
-									}
-
-									err = deleteARN(ctx, awsSession, parsed, filter, arnLogger)
-									if err != nil {
-										tracker.suppressWarning(arnString, err, arnLogger)
-										err = errors.Wrapf(err, "deleting %s", arnString)
-										continue
-									}
-									deleted[arnString] = exists
-								}
-							}
-
-							return !lastPage
-						},
-					)
-					if err != nil {
-						err = errors.Wrap(err, "get tagged resources")
-						o.Logger.Info(err)
-						matched = true
-						loopError = err
-					}
-				}
-
-				if matched {
-					nextTagClients = append(nextTagClients, tagClient)
-				} else {
-					o.Logger.Debugf("no deletions from %s, removing client", tagClientNames[tagClient])
+			instancesToDelete, err := o.findEC2Instances(ctx, ec2Client, deleted)
```

Comment:
I would like to add a comment about why EC2 instances are special and need to be in the terminated state first before attempting to delete other resources. Can someone shed light on this for me?

Reply:
This commit message should provide context on why we stop instances before anything else. https://github.com/openshift/installer/commit/cf69c1e52a012909c0ecf69ebeb6ccc6ccd8cd3f

Reply:
Great! Thanks.